### PR TITLE
Add image reader

### PR DESCRIPTION
**Pull request by apertovs**

### CODE REVIEW

**Comment by apertovs** [comment]
@KSTconsult, @RodionovF, @TolyaTalamanov, @l-bat Have a look please

File: src/io/image_reader.cpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    for (int y = 0; y < height; y++) {
+        free(row_pointers[y]);
+    }
+    free(row_pointers);
+    fclose(infile);
+    png_destroy_read_struct(&png_ptr, &info_ptr, NULL);
+    return out_tensor;
+#else
+    throw std::runtime_error("Couldn't find LIBPNG");
```
Comment (TolyaTalamanov) [inline src/io/image_reader.cpp:143]: `DeepWorks_Assert`

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -30,4 +33,21 @@ if (WITH_EIGEN)
...
+    target_compile_definitions(${PROJECT_NAME} PRIVATE HAVE_PNG)
+endif(WITH_PNG)
+
+if (WITH_JPEG)
+    find_package(JPEG REQUIRED)
+    target_link_libraries(${PROJECT_NAME} PRIVATE JPEG::JPEG)
+    target_compile_definitions(${PROJECT_NAME} PRIVATE HAVE_JPEG)
+endif(WITH_JPEG)
+
+if (WITH_EIGEN)
```
Comment (TolyaTalamanov) [inline src/CMakeLists.txt:48]: It already exists, check above

File: src/io/image_reader.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    (void) jpeg_finish_decompress(&cinfo);
+    jpeg_destroy_decompress(&cinfo);
+    fclose(infile);
+    return out_tensor;
+#else
+    DeepWorks_Assert(false && "Couldn't find LIBJPEG");
+#endif
+    return {};
```
Comment (TolyaTalamanov) [inline src/io/image_reader.cpp:75]: Why it isn't in `else` branch

File: src/io/image_reader.cpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+    png_read_info(png_ptr, info_ptr);
+
+    int width = static_cast<int>(png_get_image_width(png_ptr, info_ptr));
+    int height = static_cast<int>(png_get_image_height(png_ptr, info_ptr));
+    int channels = static_cast<int>(png_get_channels(png_ptr, info_ptr));
+
+    png_read_update_info(png_ptr, info_ptr);
+
+    deepworks::Tensor out_tensor(deepworks::Shape{channels, height, width});
```
Comment (TolyaTalamanov) [inline src/io/image_reader.cpp:110]: Why not `HWC`, the image is interleaved , isn't it ?

File: tests/unit/CMakeLists.txt [hunk 5]
```diff
@@ -5,6 +5,7 @@ include_directories(${gtest_SOURCE_DIR}/include ${gtest_SOURCE_DIR})
 file(GLOB TEST_SRC_FILES ${PROJECT_SOURCE_DIR}/tests/unit/*.cpp)
 
 add_executable(${TEST_NAME} ${TEST_SRC_FILES})
+add_definitions(-DTEST_DATA_PATH=\"${PROJECT_SOURCE_DIR}/tests/testdata\")
 
 target_link_libraries(${TEST_NAME} gtest gtest_main)
-target_link_libraries(${TEST_NAME} ${PROJECT_NAME})
+target_link_libraries(${TEST_NAME} ${PROJECT_NAME})
```
Comment (TolyaTalamanov) [inline tests/unit/CMakeLists.txt:11]: remove new line

**Comment by TolyaTalamanov** [comment]
Please, pull the latest `main`

File: src/io/image_reader.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        std::stringstream fmt;
+        fmt << "File is not a PNG file: " << path;
+        DeepWorks_Assert(false && fmt.str().c_str());
+    }
+    png_structp png_ptr = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+
+    png_infop info_ptr = png_create_info_struct(png_ptr);
+    if (setjmp(png_jmpbuf(png_ptr))) {
+        DeepWorks_Assert(false && "Error during init_io");
+    }
```
Comment (l-bat) [inline src/io/image_reader.cpp:98]: `DeepWorks_Assert(setjmp(png_jmpbuf(png_ptr)) && "Error during init_io");`

File: src/io/image_reader.cpp [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    png_read_update_info(png_ptr, info_ptr);
+
+    deepworks::Tensor out_tensor(deepworks::Shape{channels, height, width});
+
+    deepworks::Tensor::Type *dst_data = out_tensor.data();
+    deepworks::Strides tensor_strides = out_tensor.strides();
+    /* read file */
+    if (setjmp(png_jmpbuf(png_ptr))) {
+        DeepWorks_Assert(false && "Error during read image");
+    }
```
Comment (l-bat) [inline src/io/image_reader.cpp:117]: `DeepWorks_Assert(setjmp(png_jmpbuf(png_ptr)) && "Error during read image");`

File: tests/unit/CMakeLists.txt (hunk 5 above)
Comment (l-bat) [inline tests/unit/CMakeLists.txt:11]: Please don't change this line

File: src/io/image_reader.cpp [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    (void) jpeg_start_decompress(&cinfo);
+
+    size_t row_stride = cinfo.output_width * cinfo.output_components;
+    JSAMPARRAY buffer = (*cinfo.mem->alloc_sarray)((j_common_ptr) &cinfo, JPOOL_IMAGE, row_stride, 1);
+
+    int width = static_cast<int>(cinfo.output_width);
+    int height = static_cast<int>(cinfo.output_height);
+    int channels = static_cast<int>(cinfo.output_components);
+
+    deepworks::Tensor out_tensor(deepworks::Shape{height, width, channels});
```
Comment (TolyaTalamanov) [inline src/io/image_reader.cpp:49]: `copy_n` would be enough

Comment (apertovs) [reply src/io/image_reader.cpp:51]: Done

File: src/io/image_reader.cpp (hunk 7 above)
Comment (apertovs) [reply src/io/image_reader.cpp]: Done

File: src/io/image_reader.cpp (hunk 6 above)
Comment (apertovs) [reply src/io/image_reader.cpp]: Done

File: tests/unit/CMakeLists.txt (hunk 5 above)
Comment (apertovs) [reply tests/unit/CMakeLists.txt]: Done

File: src/io/image_reader.cpp (hunk 4 above)
Comment (apertovs) [reply src/io/image_reader.cpp]: Changed

File: src/io/image_reader.cpp (hunk 1 above)
Comment (apertovs) [reply src/io/image_reader.cpp]: Done

File: tests/unit/CMakeLists.txt [hunk 9]
```diff
@@ -5,6 +5,13 @@ include_directories(${gtest_SOURCE_DIR}/include ${gtest_SOURCE_DIR})
 file(GLOB TEST_SRC_FILES ${PROJECT_SOURCE_DIR}/tests/unit/*.cpp)
 
 add_executable(${TEST_NAME} ${TEST_SRC_FILES})
+add_definitions(-DTEST_DATA_PATH=\"${PROJECT_TESTS_PATH}/testdata\")
 
 target_link_libraries(${TEST_NAME} gtest gtest_main)
 target_link_libraries(${TEST_NAME} ${PROJECT_NAME})
+if (ENABLE_TESTS_ADDRESS_SANITIZER)
```
[the remark this answers is not in the document]
Comment (apertovs) [reply tests/unit/CMakeLists.txt]: Done

**Comment by TolyaTalamanov** [comment]
Great, thanks!

**Pull request merged**